### PR TITLE
Generate unique OSC resources per machine image name

### PR DESCRIPTION
**Pull request by stoyanr**
**How to categorize this PR?**

/area quality, ops-productivity
/kind bug

**What this PR does / why we need it**:
Causes `gardenlet` to generate unique `OperatingSystemConfig` resources per machine image name. This is needed because the machine image name is the `spec.type` of the OSC resource, and this field is immutable, similarly to `spec.type` fields of all other extension resources. With the `DenyInvalidExtensionResources` feature gate enabled (see #4499), attempting to change the `machine.image.name` field of one of the workers of an existing shoot would result in validation errors similar to:

```
"{\"level\":\"info\",\"ts\":\"2021-09-14T09:12:21.536Z\",\"logger\":\"extension_resources\",\"msg\":\"Invalid extension resource detected\",\"operation\":\"UPDATE\",\"error\":\"OperatingSystemConfig.extensions.gardener.cloud \"shoot--core--d066080sd2/cloud-config-worker-uq358-c8518-original\" is invalid: spec.type: Invalid value: \"ubuntu\": field is immutable\"}\n"
```

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
Another option for addressing this issue would have been to change the validation in order to no longer treat the `spec.type` field as immutable. This would work (if the feature gate is disabled, it actually works by updating the existing OSC resource). However, this would be inconsistent with all other extension resources, and therefore the solution proposed here should be preferred.

**Release note**:

```other operator
gardenlet now generates unique OSC resources per machine image name in order to avoid validation errors.
```

### CODE REVIEW

**Comment by vpnachev** [comment]
Will this change trigger machines roll out?
How is it ensured the existing nodes will start watching the new cloud-config secrets?

**Review by timuthy — COMMENTED** [review]
+1 to what @vpnachev commented.
If I saw it correctly, the OSC name does not have an effect on the machine rollout, i.e. the transition will be smooth and the "old" OSC is cleaned up by `Delete stale operating system config resources` in the shoot flow. But it would be nice if you can double check this.

**Comment by stoyanr** [comment]
> Will this change trigger machines roll out?

Changing the `machine.image.name` does trigger a machine rollout of course. But reconciling an old shoot that was created without this change doesn't, and the shoot reconciliation is stuck at 98% on the step "Waiting until all shoot worker nodes have updated the cloud config user data".

@timuthy @vpnachev Do you have a good idea what to do here? Should we trigger a (one-time) node rollout, and what's the best way to do it?

/hold

**Comment by timuthy** [comment]
> Should we trigger a (one-time) node rollout, and what's the best way to do it?

We should not do that as this causes a huge disruption. I don't have an ad-hoc idea, let me get back later.

**Comment by stoyanr** [comment]
> We should not do that as this causes a huge disruption. I don't have an ad-hoc idea, let me get back later.

In fact, I realized that what we need here is just to change the name of the OSC resource, not the name of the cloud config secret. Therefore I reduced the scope of the change. I believe this addresses the above concern, no node rollout is needed, etc.

I decided to append the machine image name as it is to make it more explicit, if you have any concerns regarding length I could append a SHA256 hash instead.

I tested successfully:
* Creating a shoot without the change and reconciling it with the change, and vice versa. 
* Changing the machine image name of a worker (a few times).

If you still have concerns or ideas what else to test, please let me know.

/unhold

**Comment by stoyanr** [comment]
@timebertt @vpnachev From my perspective this could be included in v1.32?
/milestone v1.32

**Review by timuthy — APPROVED** [review]
/lgtm
Also tested it successfully, thanks for the PR @stoyanr.
FYI: This should be coordinated with https://github.com/gardener/gardener/pull/4674 (/cc @timebertt)

**Comment by vpnachev** [comment]
Is this also restricted to only when CachedRuntimeClient feature gate is enabled?
/unmilestone v1.32
until the scope of the change is clarified.

**Comment by vpnachev** [comment]
Actually, does not matter. It is for sure related to another alpha feature gate `DenyInvalidExtensionResources`.

**Comment by stoyanr** [comment]
> Actually, does not matter. It is for sure related to another alpha feature gate `DenyInvalidExtensionResources`.

@vpnachev It is a prerequisite for successfully enabling this feature gate. Without this change, if the feature gate is enabled attempting to change the machine image name of a worker pool will result in validation errors. However, the change itself does not depend in any way on this feature gate.

Why is this a problem for including this in v1.32?

**Comment by timebertt** [comment]
> FYI: This should be coordinated with #4674 (/cc @timebertt)

Feel free to merge this one before #4674. My PR should be easy to rebase, once this one is it.

**Comment by voelzmo** [comment]
I might not entirely understand the flow after this change is merged, maybe you can help me if this is correct:
* Until https://github.com/gardener/gardener/blob/f7e980d9e5a34524ffb305806e13cf4b2bc6b692/pkg/operation/botanist/component/extensions/operatingsystemconfig/operatingsystemconfig.go#L317 is executed, two OSCs with different contents exist
* Both OSCs are reconciled into the same secret, as the secret key didn't change
* The cloud-config-downloader gets and applies the secret content every regularly (I think every minute?)

Can it happen that a worker pulls down the secret with the old content somehow? Or is the flow configuration in https://github.com/gardener/gardener/blob/f7e980d9e5a34524ffb305806e13cf4b2bc6b692/pkg/gardenlet/controller/shoot/shoot_control_reconcile.go#L321-L335 taking care of this never happening?

**Comment by stoyanr** [comment]
@voelzmo 

> is executed, two OSCs with different contents exist

In practice, `DeleteStaleResources` is invoked very quickly after the new OSCs have been deployed, even without waiting for them to be reconciled. During testing, I never actually saw both OSCs at the same time (although they should have existed for some parts of the second of course).

> Can it happen that a worker pulls down the secret with the old content somehow?

* There is no "old content" if nothing else changed (an existing shoot is reconciled with the change for the first time), the secret content doesn't change as well.
* In case something else changed (e.g. the machine image name), the secret name is still the same, and whatever mechanisms worked before are still in place.

The change affects only the name of the OSC resource, not the name or the content of the secret. It is only a "superficial" change to avoid validation errors, and it should not change any existing mechanisms as far as I can tell.

**Pull request merged**
